### ISSUE TITLE
#pragma support (OpenMP specifically)

### ISSUE THREAD

**Issue by alashworth** [alashworth/stan-monorepo-stage1#16]
<a href="https://github.com/bgoodri"><img src="https://avatars2.githubusercontent.com/u/3386282?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [bgoodri](https://github.com/bgoodri)**
_Monday Dec 09, 2013 at 05:28 GMT_
_Originally opened as https://github.com/stan-dev/stan/issues/439_

----

1. Amend the parser to not throw away anything to the right (inclusive) of `#pragma ...`
2. There is no step 2.

**Comment by alashworth** [comment]
<a href="https://github.com/bgoodri"><img src="https://avatars2.githubusercontent.com/u/3386282?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bgoodri](https://github.com/bgoodri)**
_Tuesday Dec 10, 2013 at 16:56 GMT_

----

For some reason, this issue is not showing up in the issues list on GitHub, so mentioning @bob-carpenter to make sure it does not get lost

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Tuesday Dec 10, 2013 at 19:35 GMT_

----

I'm responding through GitHub to let you know it is showing up.  At:

https://github.com/stan-dev/stan/issues/439

**Comment by alashworth** [comment]
<a href="https://github.com/bgoodri"><img src="https://avatars2.githubusercontent.com/u/3386282?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bgoodri](https://github.com/bgoodri)**
_Sunday Apr 20, 2014 at 23:51 GMT_

----

With g++-4.9 scheduled to be released on Tuesday and it supporting OpenMP 4.0, it would be nice if Stan could support pragmas. As far as I can tell, all that is needed is
1. For the parser to interpret #pragma literally rather than a R-style comment
2. For the parser to include such lines in the .cpp file
3. Add a Makefile option to add `-include omp.h -fopenmp` to the compile statement

I know how to do 3, so that gets us halfway there.

**Comment by alashworth** [comment]
<a href="https://github.com/mbrubake"><img src="https://avatars0.githubusercontent.com/u/3383416?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mbrubake](https://github.com/mbrubake)**
_Sunday Apr 20, 2014 at 23:56 GMT_

----

Wouldn't the autodiff stack need to be made more threadsafe and efficient
before this was a good idea?

**Comment by alashworth** [comment]
<a href="https://github.com/bgoodri"><img src="https://avatars2.githubusercontent.com/u/3386282?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bgoodri](https://github.com/bgoodri)**
_Monday Apr 21, 2014 at 00:10 GMT_

----

You may be right about loops that include sampling statements (that people
should vectorize anyway) like

#pragma omp parallel for

for(i in 1:N) y[i] ~ normal(0,1);

but the current generated .cpp files accumulate the statements that change
the log-posterior until the end of the model block and evaluate them
(serially) with

return lp_accum__.sum();

So, I don't think parallelizing the loop is thread unsafe even in that
case. But if it is, then we can say to only parallelize pure assignment
loops.

Ben

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Monday Apr 21, 2014 at 11:28 GMT_

----

Michael --- you should check this out because it's related
to the new auto-diff.

More inline.
- Bob

On Apr 21, 2014, at 2:10 AM, bgoodri notifications@github.com wrote:

> You may be right about loops that include sampling statements (that people 
> should vectorize anyway) like 
> 
> #pragma omp parallel for 
> 
> for(i in 1:N) y[i] ~ normal(0,1); 

Correct --- these can't be executed in parallel because they modify
the shared auto-diff stack.  

> but the current generated .cpp files accumulate the statements that change 
> the log-posterior until the end of the model block and evaluate them 
> (serially) with 
> 
> return lp_accum__.sum(); 

What gets accumulated is the statement values --- the statements themselves
are evaluated eagerly as they are found. 

> So, I don't think parallelizing the loop is thread unsafe even in that 
> case. But if it is, then we can say to only parallelize pure assignment 
> loops. 

When you get to .sum(), it's not changing the auto-diff stack, just
doing a bunch of operations on doubles.  I'm not sure they'll be
easy to parallelize given the indirection through vari, but it's safe.  
I'd think the new approach Michael's working on should have better vectorization 
properties, but it's always hard to tell --- we were wrong about first order 
being faster for reasons we still don't quite understand.

Also, when we start doing things like Jacobian calculations and Hessians,
there are embarassingly parallel optimization opportunities in the way we're
doing things now that won't be so easy the way Michael's reformulating things
with a reusable autodiff stack.  
- Bob

> Ben 
> 
> On Sun, Apr 20, 2014 at 7:56 PM, Marcus Brubaker

**Comment by alashworth** [comment]
<a href="https://github.com/betanalpha"><img src="https://avatars0.githubusercontent.com/u/1527190?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [betanalpha](https://github.com/betanalpha)**
_Monday Apr 21, 2014 at 12:52 GMT_

----

> Also, when we start doing things like Jacobian calculations and Hessians, 
> there are embarassingly parallel optimization opportunities in the way we're 
> doing things now that won't be so easy the way Michael's reformulating things 
> with a reusable autodiff stack. 

Actually this won’t be so bad given that much of the stack is read-only so we
can still do some embarrassing parallelization (oh, how I hate that name by
the way).

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Monday Apr 21, 2014 at 13:21 GMT_

----

On Apr 21, 2014, at 2:52 PM, Michael Betancourt notifications@github.com wrote:

> > Also, when we start doing things like Jacobian calculations and Hessians, 
> > there are embarassingly parallel optimization opportunities in the way we're 
> > doing things now that won't be so easy the way Michael's reformulating things 
> > with a reusable autodiff stack. 
> 
> Actually this won’t be so bad given that much of the stack is read-only so we 
> can still do some embarrassing parallelization 

That's a good point.  Rather than having a single shared stack
for each row of the Jacobian, new ones could be created in different
threads.  So you'd have a pass to build up the partials, then multiple
passes in parallel to do calcs, each with their own workspace.

This is kind of like what Adept did in creating the adjoint stack only
at the end when needed.

> (oh, how I hate that name by 
> the way).

Agreed on the name.  It's wrong because it's a good thing in practice --- it's
just wrong for a paper on parallel processing.
- Bob

**Comment by alashworth** [comment]
<a href="https://github.com/syclik"><img src="https://avatars2.githubusercontent.com/u/425751?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [syclik](https://github.com/syclik)**
_Wednesday Nov 30, 2016 at 14:42 GMT_

----

@bgoodri, is there a concrete suggestion you have? Is this something we need to code generate? Is this something that should go down in the math library? Maybe in the distributions?

**Comment by alashworth** [comment]
<a href="https://github.com/bgoodri"><img src="https://avatars2.githubusercontent.com/u/3386282?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bgoodri](https://github.com/bgoodri)**
_Wednesday Nov 30, 2016 at 14:58 GMT_

----

The first was "don't drop comments that start with a #", in which case users could do their own #pragmas. But users would still have to add `-include omp.h -fopenmp` to get it to work, and could crash their program (which I don't care much about) if they try to do something in parallel that has to be serial. The second was to introduce a `pfor (j in 1:J)` block that would code generate to 
```
#pragma omp for
for (j in 1:J) {
  // do stuff that refers to j
}
```
in which case we would at least know to generate a line with `#include omp.h` and possibly we could do the for loop in a safer way. Sebastian was playing around with different `#pragmas` that would preclude some problems in Stan.

I doubt there is any distribution in Stan Math with accumulated analytical gradients where OpenMP would be worth the overhead. And Bob is suggesting that distributions without accumulated analytical gradients would be problematic for OpenMP. Eigen is doing some stuff like this for operations on giant matrices.

Also, none of the OpenMP stuff works on a Mac with Xcode. You have to install clang or g++ from upstream.

**Comment by alashworth** [comment]
<a href="https://github.com/syclik"><img src="https://avatars2.githubusercontent.com/u/425751?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [syclik](https://github.com/syclik)**
_Wednesday Nov 30, 2016 at 15:13 GMT_

----

Let's start with the first thing you mentioned: "don't drop comments that start with a #". Can you tell me if I'm right in my interpretation?
- deprecate `#` as a comment character
- directly copy lines that start with `#` into the generated C++ code
If so, can you create a new issue that clearly states that? If I'm mistaken, can you correct me?

Regarding OpenMP, @rtrangucci and I actually got a version of clang++ installed (I forget how now, but it's not Xcode) and showed a little speed boost on a super-large matrix multiplication, but that's neither here nor there.

Should we close this issue and create two issues:
1. something to do with `#` as directives and not comments
2. creating a `pfor` function in the Stan language?

**Comment by alashworth** [comment]
<a href="https://github.com/bgoodri"><img src="https://avatars2.githubusercontent.com/u/3386282?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bgoodri](https://github.com/bgoodri)**
_Wednesday Nov 30, 2016 at 16:05 GMT_

----

> If so, can you create a new issue that clearly states that?

That is literally the OP on this issue, but it is an open question as to whether that is a good idea. The `pfor` thing could be a separate issue, but we should probably do one or the other rather than both.

**Comment by alashworth** [comment]
<a href="https://github.com/syclik"><img src="https://avatars2.githubusercontent.com/u/425751?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [syclik](https://github.com/syclik)**
_Wednesday Nov 30, 2016 at 16:27 GMT_

----

The OP wasn't clear enough for me.

If you think we should do one and not both, please help us make that decision. What are the pros and cons of doing one or the other? What's your preference? Who's actually going to implement it and what's that person's preference?

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Wednesday Nov 30, 2016 at 18:56 GMT_

----

I suggest we also move this to longer-term issues
until there's more of a design as to what we should do.

We don't need to decide who's going to implement something
when we create an issue---it just has to be precise enough
to be implemented.

For the OpenMP pragma, I don't think there's any way we
could parallelize Stan code naively without synchronizing
on the expression graph for autodiff.  That would presumably
require the pieces that are being executed in parallel to
not touch the expression graph, which means analytical derivatives.

The case Sebastian was demo-ing ran multiple ODE solvers
in parallel.  And I think he had a way to synch the results
(maybe another comment-like thing?)
